### DISCUSSION TITLE
what's the best environment for video playing?

### DISCUSSION THREAD

**Discussion by H-MLim** [umlaeute/Gem#397] [Q&A]
hi. I have been suffering the video playback problem on Windows since many years ago. (please see #262 ). if you are Windows users, and having no problem for playing video with `pix_film` or `pix_movie`, please let me know your video format and encoding settings. (or also, let me know if you have a same problem) many thanks.

**Comment by umlaeute** [comment]
not on Windows, but so far the best performance i've seen on Linux was with using `[pix_video]` with the `videoVLC` backend (which can play back movie files as if it were a video, using the VLC-engine for decoding; although it eats much more CPU than a native VLC playback, [4k movies](http://distribution.bbb3d.renderfarming.net/video/mp4/bbb_sunflower_2160p_30fps_normal.mp4) play back at "normal speed".

the obvious show-stopper is, that currently the Gem builds for Windows do not include the `videoVLC` backend (and it used to crash on W32, as mentioned in https://github.com/umlaeute/Gem/issues/64).

so not an answer at all, just a data point :-)

**Comment by kubriel** [comment]
Hallo, i work a lot with video playback in Gem on old machines. Theese things are depending on your platform, so i see it from linux perspective. In bugreport you were talking about mov or mp4, but thats only containers. My feeling is that they dont have big impact on performance. Inside container you have some codec, and that's more important. One of the most popupar is mjpeg. Its very fast also if you need to access your frames in random order. But the file will be a bit big, so you have to guard the data rate, if your storage is fast enough. pix_film and pix_movie are decoding video in separate thread, so i was often cutting video in 4 quarters and playing them synchrous with four pix_movie or pix_film objects to a framebuffer, so i can use it as one texture. This way you can make better use of your multicore processor.
Other option might be to use pix_video with VLC plugin. This way you might be able to use hardware decoding on your GPU, but i was never really happy with this.
So last years i use mostly Avilleret's gem-mpv:
https://github.com/avilleret/gem-mpv
I was able to nicely playback 15500x850(i dont remember numbers exactly) 30fps on an ivy bridge i7, thats a blast! Or not? It's a gemframebuffer interface to mpv media player, and i found this solution very efficient to decode h264, or HAP codecs. Thing is that gem-mpv seems abandoned, and mpv api changed, so i forked it and adapted it to new mpv's render api, but i think it still doesn't work with libmpv2, so maybe its not very usefull. 
https://github.com/kubriel/pd-mpv
Anyway that's still software decoding, but hardware decoding with that is very close i think. I was able to run it with vaapi-copy, but that will copy frames back to cpu, and then again back to gpu. I just dont know how to pass display resource to mpv parameters. Its my dream to have that!

**Reply by H-MLim** [reply]
hi. thanks for the info. that sounds so intriguing! 
but since I'm totally newbie for that, I'm afraid I don't know how to install that `gem-mpv` (and `libmpv`) libraries into my Pd. can I install it on my Windows too? (I mean, is `mpv.pd_linux` only for linux?)

by the way fortunately, I don't have any problems on `pix_video` (with webcam). just for `film` and `movie`.

**Reply by kubriel** [reply]
I have no idea about windows, but really try this mjpeg, should run nicely in mov container.
